lib/DXIL/DxilModule.cpp

```diff
@@ -2081,4 +2111,157 @@ bool DxilModule::IsPrecise(const Instruction *inst) const {
+      // update min shader model and shader stage mask per function
+      UpdateFunctionToShaderCompat(&function);
+    } else if (!function.isDeclaration()) {
+      // Initialize worklist with functions with callers.
+      // only used for validator version 1.8+
+      if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) >= 0 &&
+          !function.user_empty())
+        worklist.insert(&function);
+
+      // Skip functions with shader flags already set
+      if (m_FuncToShaderCompat.count(&function) != 0)
+        continue;
+
+      // Collect shader flags for function
+      ShaderCompatInfo &info = m_FuncToShaderCompat[&function];
+      info.shaderFlags = ShaderFlags::CollectShaderFlags(&function, this);
+    }
+  }
+
+  // Propagate ShaderCompatInfo to callers, limit to 1.8+ for compatibility
+  if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) >= 0) {
+    while (!worklist.empty()) {
+      llvm::Function *F = worklist.pop_back_val();
+      ShaderCompatInfo &calleeInfo = m_FuncToShaderCompat[F];
+      // Update callers
+      for (auto U : F->users()) {
+        if (CallInst *CI = dyn_cast<CallInst>(U)) {
+          llvm::Function *caller = CI->getParent()->getParent();
+          // Merge info, if changed and called, add to worklist so we update
+          // any callers of caller as well.
+          if (m_FuncToShaderCompat[caller].Merge(calleeInfo) &&
+              !caller->user_empty())
+            worklist.insert(caller);
+        }
+      }
+    }
+  }
+
+  // We must select the appropriate shader mask for the validator version,
+  // so we don't set any bits the validator doesn't recognize.
+  unsigned ValidShaderMask =
+      (1 << ((unsigned)DXIL::ShaderKind::LastValid + 1)) - 1;
+  if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 5) < 0) {
+    ValidShaderMask = (1 << ((unsigned)DXIL::ShaderKind::Last_1_4 + 1)) - 1;
+  } else if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) < 0) {
+    ValidShaderMask = (1 << ((unsigned)DXIL::ShaderKind::Last_1_7 + 1)) - 1;
+  } else if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) >= 0) {
+    ValidShaderMask = (1 << ((unsigned)DXIL::ShaderKind::Last_1_8 + 1)) - 1;
+  }
+
+  for (auto &function : GetModule()->getFunctionList()) {
+    if (!function.isDeclaration()) {
+      ShaderCompatInfo &info = m_FuncToShaderCompat[&function];
+      DXIL::ShaderKind shaderKind = DXIL::ShaderKind::Library;
+      const DxilFunctionProps *props = nullptr;
+      if (HasDxilFunctionProps(&function)) {
+        props = &GetDxilFunctionProps(&function);
+        shaderKind = props->shaderKind;
+      }
+
+      if (shaderKind == DXIL::ShaderKind::Library)
+        info.mask &= ValidShaderMask;
+      else
+        info.mask &= (1U << static_cast<unsigned>(shaderKind));
+
+      // Increase min target based on features used:
+      ShaderFlags &flags = info.shaderFlags;
+      if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) >= 0) {
```

Comment:
There are a few of these checks for greater than 1.8 sprinkled in this function. I can't tell quickly if this function has any effect with a version less than 1.8. Regardless, can any be consolidated for maintainability and legibility?

Reply:
It does have an effect for < 1.8.  The 1.8 checks are to make sure we generate the same result in validator 1.7 and below.

The effects when < 1.8 are:
- CollectShaderFlags for each function
- Updates each functions min target and valid shader stage flags based on intrinsic usage via UpdateFunctionToShaderCompat on intrinsics to update users.
- Masks shader stage flags based on ShaderKind for a shader entry function, or based on stages that exist in the shader model for a library function.
- Updates min target based on use of a few optional feature flags (6.2 for native low precision, 6.1 for barycentrics or ViewID)

With >= 1.8, we also:
- propagate flags from internal calees to callers
- use AdjustMinimumShaderModelAndFlags to account for all feature flags, shader stages, WaveSize use, and handle the special case for derivative use.